### PR TITLE
FEATURE: Thread-safe MetricLogger

### PR DESCRIPTION
**Pull request by coyotebush**
Use an AtomicBoolean for the shutdown flag, and remove backing queues
other than BlockingQueue.

For clients using only the `MetricLoggerLike` interface, this should suffice.

### CODE REVIEW

File: balboa-client/src/main/scala/com/socrata/metrics/collection/SocrataQueues.scala [hunk 1]
```diff
@@ -2,8 +2,6 @@ package com.socrata.metrics.collection
 
 import java.util.concurrent.{LinkedBlockingQueue, TimeUnit}
 
-import scala.collection.mutable.Queue
-
 trait SocrataAbstractQueue[Q] {
```
Comment (coyotebush) [inline balboa-client/src/main/scala/com/socrata/metrics/collection/SocrataQueues.scala]: We could just take out SocrataAbstractQueue and have PreBufferQueue store a BlockingQueue directly.

File: balboa-client/src/main/scala/com/socrata/metrics/collection/PreBufferQueues.scala [hunk 2]
```diff
@@ -1,13 +1,13 @@
 package com.socrata.metrics.collection
 
-import java.util.concurrent.LinkedBlockingQueue
+import java.util.concurrent.{BlockingQueue, LinkedBlockingQueue}
 
 import com.socrata.metrics.components.MetricEntry
 
 trait PreBufferQueue {
-  def queue:SocrataAbstractQueue[MetricEntry]
+  def queue: BlockingQueue[MetricEntry]
```
Comment (mhotan) [inline balboa-client/src/main/scala/com/socrata/metrics/collection/PreBufferQueues.scala]: Yes! Way to stop us from reinventing the wheel :-)

File: balboa-client/src/main/scala/com/socrata/metrics/collection/SocrataQueues.scala [hunk 3]
```diff
@@ -1,29 +0,0 @@
-package com.socrata.metrics.collection
```
Comment (mhotan) [inline balboa-client/src/main/scala/com/socrata/metrics/collection/SocrataQueues.scala]: Yes. Love it.

File: balboa-client/src/main/scala/com/socrata/metrics/collection/PreBufferQueues.scala (hunk 2 above)
Comment (doomcrata) [inline balboa-client/src/main/scala/com/socrata/metrics/collection/PreBufferQueues.scala]: I prefer a square wheel, sir. It works better on square-ish worlds.

Comment (mhotan) [inline balboa-client/src/main/scala/com/socrata/metrics/collection/PreBufferQueues.scala]: As long as its a Socrata Square wheel SON!

File: balboa-client-jms/src/test/scala/com/socrata/balboa/impl/LoggerTests.scala [hunk 4]
```diff
@@ -56,7 +56,7 @@ class MetricLoggerSpec extends WordSpec with ShouldMatchers with MetricLoggerToQ
         with HashMapBufferComponent
         with TestMessageQueueComponent
         with LinkedBlockingPreBufferQueue
-      quickLogger.enqueue(new MetricEntry("ayn", "num_kitties", 4, 1L, aggType))
+      quickLogger.logMetric("ayn", "num_kitties", 4, 1L, aggType)
       quickLogger.stop()
       Thread.sleep(300)
```
Comment (mhotan) [inline balboa-client-jms/src/test/scala/com/socrata/balboa/impl/LoggerTests.scala]: :sob:

Comment (mhotan) [inline balboa-client-jms/src/test/scala/com/socrata/balboa/impl/LoggerTests.scala]: Nothing you have to deal with but still.... :sob:

File: balboa-client/src/main/scala/com/socrata/metrics/collection/PreBufferQueues.scala (hunk 2 above)
Comment (doomcrata) [inline balboa-client/src/main/scala/com/socrata/metrics/collection/PreBufferQueues.scala]: Typesafe is no way to go through life, SON!

Comment (housepage) [inline balboa-client/src/main/scala/com/socrata/metrics/collection/PreBufferQueues.scala]: Square-ish worlds would make the square whee only slightly less suboptimal. Now if you had a world where the vector of the gravitational acceleration was changing 90 degrees at the proper frequency, perhaps you'd have a point. Even then, the performance of square wheel vs. circular wheel would be about the same.

File: balboa-client-jms/src/test/scala/com/socrata/balboa/impl/LoggerTests.scala [hunk 5]
```diff
@@ -56,7 +56,7 @@ class MetricLoggerSpec extends WordSpec with ShouldMatchers with MetricLoggerToQ
         with HashMapBufferComponent
         with TestMessageQueueComponent
         with LinkedBlockingPreBufferQueue
-      quickLogger.enqueue(new MetricEntry("ayn", "num_kitties", 4, 1L, aggType))
+      quickLogger.logMetric("ayn", "num_kitties", 4, 1L, aggType)
```
Comment (housepage) [inline balboa-client-jms/src/test/scala/com/socrata/balboa/impl/LoggerTests.scala]: @aynleslie Kitties make everything better :-P

File: balboa-client/src/main/scala/com/socrata/metrics/collection/PreBufferQueues.scala (hunk 2 above)
Comment (jasonij) [inline balboa-client/src/main/scala/com/socrata/metrics/collection/PreBufferQueues.scala]: I wish gravity were more like the weather. Sigh.

**Comment by housepage** [comment]
Let them eat :cake:

**Comment by mhotan** [comment]
Thank you for the updates to make Metric Logger more thread safe.  I understand that with the level of complexity we introduced earlier makes it difficult to test.  Balboa must go through a serious rewrite but unfortunately now is not the right time or place.  The changes look good lets :shipit: while :two_men_holding_hands: and keep an eye out for any other concurrency issues.

File: balboa-client/src/main/scala/com/socrata/metrics/collection/PreBufferQueues.scala (hunk 2 above)
Comment (splittingfield) [inline balboa-client/src/main/scala/com/socrata/metrics/collection/PreBufferQueues.scala]: Uhm.  This was abstracted for a reason and I remember a good one.  There was no wheel reinvention, we just wanted to have a shared base class for different Queues....

Comment (coyotebush) [inline balboa-client/src/main/scala/com/socrata/metrics/collection/PreBufferQueues.scala]: Happy to hear a reason, but we weren't even using the SocrataQueue class (the one that wrapped a scala Queue).

File: balboa-client/src/main/scala/com/socrata/balboa/impl/MetricDequeuer.scala [hunk 6]
```diff
@@ -27,7 +27,7 @@ trait MetricDequeuerService {
 
       def take() {
         // if need be, we can add a cushion to prevent context switching, e.g. only take if (queue.size > 10)
-        queue.take() match {
+        Option(queue.poll(1, TimeUnit.SECONDS)) match {
```
Comment (coyotebush) [inline balboa-client/src/main/scala/com/socrata/balboa/impl/MetricDequeuer.scala]: Changed this back to `match` style which is a little clearer.

File: balboa-client/src/main/scala/com/socrata/metrics/collection/PreBufferQueues.scala (hunk 2 above)
Comment (coyotebush) [inline balboa-client/src/main/scala/com/socrata/metrics/collection/PreBufferQueues.scala]: Square wheels are actually completely appropriate in [the right circumstances](https://www.youtube.com/watch?v=LgbWu8zJubo). What we had here was an unnecessary gearbox.

**Pull request merged**